jupyter_server/extension/application.py

```diff
@@ -174,7 +174,10 @@ def _default_open_browser(self):
 
     @classmethod
     def get_extension_package(cls):
-        return cls.__module__.split('.')[0]
+        parts = cls.__module__.split('.')
+        if '(namespace)' not in str(__import__(parts[0])):
+            return parts[0]
+        return '.'.join(parts[0:2])
 
```

Comment:
Can we expect `parts` to always have at least one or two items? I wonder if a try/catch is needed to handle a possible `IndexError` exception.

Reply:
This is some strange inconsistent behaviour in Python. When you use a `:` in the index (array splicing) Python doesn't raise an `IndexError`: 

```python
>>> [][20]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
IndexError: list index out of range
>>> [][20:50]
[]
```

Which means the `join` will work no matter how many parts are in the list:

```python
>>> '.'.join([])
''
>>> '.'.join(['a'])
'a'
>>> '.'.join(['a', 'b'])
'a.b'
>>> '.'.join(['a', 'b', 'c'])
'a.b.c'
```

Reply:
I see. Thanks for clarifying! If we can expect at least 1 item, I think this is fine. 
```python
if '(namespace)' not in str(__import__(parts[0])):
```

Reply:
i think a more consistent approach, that would avoid the importing overhead would be to base the logic off of the import spec's loader. there are different loaders for namespaces, zips, and source files. the `find_spec` will return just the thing that would load the module then the logic would follow the type of the loader.

     >>> importlib.util.find_spec("wxyz").loader
     <_frozen_importlib_external._NamespaceLoader at 0x7f9fda6c0810>

Reply:
I did consider this, however, I'm not sure how portable it would be between different Python implementations.

Reply:
that is an important consideration. the importing systems have been seeing lots of changes over versions. for this use case though I think it might be safe.

i'm fairly certain that all the machinery you would need is supported for >=3.6, which i assume is lower bound constraint and we can find the [`find_spec` method in 3.6](https://github.com/python/cpython/blob/3.6/Lib/importlib/util.py#L66). the `find_spec` and `ModuleLoader` is introduced in >=3.4 according to [PEP451](https://www.python.org/dev/peps/pep-0451/).

Reply:
I was worried about the alternative Python implementations e.g. PyPy etc as I think the `_frozen_importlib_external` bits are implementation specific.

I'm not sure what the support policy is for these implementations, the Conda Forge builds are noarch so this works:

```
mamba install pypy jupyterlab
```

Reply:
Note that the current code will fail if `__file__` is empty for some other reason than it being a namespace package (`__file__` is completely optional, and might be empty is loaded from a DB https://docs.python.org/3/reference/import.html#__file__). If we are importing it anyways, why not look at the type of the `__path__` attribute instead? https://www.python.org/dev/peps/pep-0420/#differences-between-namespace-packages-and-regular-packages

Reply:
Thanks all for the input.

I've gone for a tweak of the suggestion from @tonyfast using `submodule_search_locations` as a marker (should be an instance of `_NamespacePath`). I've checked and the `_NamespacePath` is consistent between cPython and MyPy (haven't checked the other Python implementations however they don't support py3.6 so aren't jupyter_server compatible). This avoids importing the namespace, made sure it doesn't raise an error if the namespace is not importable (i.e. not installed) so error handling is not affected.

Tested against cPython3.6, cPython3.9, MyPy3.7(7.3.4) for regular and editable `pip install`s. Added unit tests.

Reply:
Thanks @oliver-sanders. I think we’re getting closer here.

While the current implementation is certainly cleaner, it depends on private APIs and classes in the standard lib. I don’t think we want to do this. Those APIs could, in principle, change at any point, without warning, and catastrophically break jupyter_server on import. If those APIs become public+stable in sometime in this future, I believe this would be the way to go.

Unfortunately, right now, I think we’re stuck with something like @vidartf suggests—importing and inspecting the package's `__path__`.

Reply:
>  it depends on private APIs and classes in the standard lib.

> suggests—importing and inspecting the package's __path__.

Unfortunately this suggested implementation is effectively the same as the current implementation just with an added import step and both approaches rely on the same private interface.

Here are a couple of example's to demonstrate. Set your `PYTHONPATH` to pick up the `namespace-package-test` dir on my branch like so:

```console
$ cd ~/jupyter_server
$ git checkout oliver-sanders/extensions-in-namespace-packages
$ PYTHONPATH="$PWD/jupyter_server/tests/namespace-package-test:$PYTHONPATH" python
```

Import the module then inspect the `__path__`  attribute:

```python
>>> import test_namespace
>>> test_namespace.__path__
_NamespacePath(['.../jupyter_server/tests/namespace-package-test/test_namespace'])
>>> import jupyter_server
>>> jupyter_server.__path__
['.../jupyter_server/jupyter_server']
```

Use importlib to determing the `__path__` attribute without importing:

```python
>>> import importlib.util
>>> importlib.util.find_spec('test_namespace').submodule_search_locations
_NamespacePath(['.../jupyter_server/tests/namespace-package-test/test_namespace'])
>>> importlib.util.find_spec('jupyter_server').submodule_search_locations
['.../jupyter_server/jupyter_server']
```

Both approaches give the same result. In order to tell between the regular and namespace package we must check the type of the path object.

We have two options for doing this:

1. Use the private `_frozen_importlib_external` interface.

   **pro:** The code will break completely if the private interface were ever to change, this will get quickly picked up by the GitHub actions tests which run for all supported Python versions / implementations.

   **con:** If cPython and PyPy were to go different ways this approach would require review.

2. Test the name of the class , [PEP420 mandates](https://www.python.org/dev/peps/pep-0420/#nested-namespace-packages) that it should be `'_NamespacePath'`.

    **pro:** Should work for all Python implementations.

     **con:** Testing a string is not a brittle as an `import/isinstance` check, if the internal representation of the class were to differ subtly the logic could break in some environments.

Happy to go either way, let me know what is preferred.

Reply:
Let's keep this as implemented. (see my comment [here](https://github.com/jupyter-server/jupyter_server/pull/523#issuecomment-874898341)).